### PR TITLE
Integrate DAI and PDAI

### LINKED ISSUE #111: Integrate DAI token and pDAI
## Scope

- Submit a governance proposal for DAI mock on the testnet and eventually mainnet
- Check and fix all Exchange.js functionality:
  - ETH <-> DAI
  - DAI <-> pDAI
- Enable fast exits for pDAI
- Might be that logic is very similar to what we've implemented with sunDAI already

## Deliverables

- pDAI in the burner-wallet

## Gain for the project

- Fast transactions and no volatility in holding crypto 💥 

## Roles
bounty gardener: @TimDaub / 10%
bounty worker: name / 75%
bounty reviewer: name / 15%

### LINKED ISSUE #138: Update config.js back to testnet.leapdao.org
Migration is complete: https://leapdao.slack.com/archives/CFKUNHHC7/p1558707971002600?thread_ts=1558707724.001800&cid=CFKUNHHC7

### PR DESCRIPTION
**Pull request by TimDaub**
Fixes #111

TODOs:

- [x] Upgrade from testnet to theta
- [x] Fix deposits DAI => PDAI (metamask)
- [x] Remove `P_DAI_TOKEN_ADDR` as it's the same as `DAI_TOKEN_ADDR`
- [x] Fix deposits DAI => PDAI (no metamask)
- [x] Adjust `getUnspents`. Breaking change in `leap-core`
- [x] Fix exits PDAI => DAI (metamask)
- [x] Fix exits PDAI => DAI (no metamask)
- [x] Moved Mainnet testing into separate issue from this PR: https://github.com/leapdao/burner-wallet/issues/137
- [x] Pin latest version of leap-core when released: https://github.com/leapdao/leap-core/pull/110
- [x] When #134 is merged, base this branch on it and fix #138 "Upgrade theta to testnet"
- [x] Sending DAI (metamask) to regular address, burner-wallet asks for approval and ultimately deposits tokens to Bridge lol
- [x] Tests (all Rinkeby):
  - [x] Deposit as PDAI (metamask)
  - [x] Deposit as PDAI (no metamask)
  - [x] Exit as DAI (metamask)
  - [x] Exit as DAI (no metamask)

Comments:

- According to @troggy exits do not work on theta yet: https://leapdao.slack.com/archives/CFKUNHHC7/p1558535288006700?thread_ts=1558534628.005300&cid=CFKUNHHC7

### CODE REVIEW

**Comment by troggy** [comment]
@TimDaub fast exits are working on the testnet now. Requires leap-core with [#110](https://github.com/leapdao/leap-core/pull/110) merged though.

**Comment by johannbarbie** [comment]
open a related issue: https://github.com/leapdao/burner-wallet/issues/154

**Comment by johannbarbie** [comment]
I'm testing the exits. I've started an exit with my address `0x8db6B632D743aef641146DC943acb64957155388`, but don't see this address neither in http://theta.leapdao.org/status nor in dynamodb. what am i doing wrong?

**Comment by TimDaub** [comment]
> but don't see this address neither in http://theta.leapdao.org/status

Shouldn't you check on https://testnet.leapdao.org/status?



> nor in dynamodb

Not sure, could also be the wallet. When I check for exits on the market maker manually here, I can't see anything either: https://2nuxsb25he.execute-api.eu-west-1.amazonaws.com/testnet/exits/0x8db6B632D743aef641146DC943acb64957155388/0

**Comment by troggy** [comment]
https://testnet.leapdao.org is the correct bridge ui

**Comment by johannbarbie** [comment]
the exit was > 1 LEAP, no-where listed. @troggy how would i start debugging the exit maker?

File: src/components/Exchange.js [hunk 1]
```diff
@@ -946,7 +967,7 @@ export default class Exchange extends React.Component {
             icon={'ArrowDownward'}
             disabled={
               buttonsDisabled ||
-              (!this.state.notSundai && this.state.exitableSunDaiBalance === 0) ||
+              (!this.state.notSundai && this.state.exitableSunDaiBalance ===0) ||
```
Comment (johannbarbie) [inline src/components/Exchange.js]:
```suggestion
              (!this.state.notSundai && this.state.exitableSunDaiBalance === 0) ||
```

File: src/components/Exchange.js [hunk 2]
```diff
@@ -874,16 +900,10 @@ export default class Exchange extends React.Component {
                   };
 
                   // TODO: get real decimals
-                  const amount = bi(exitableAmount * 10 ** 18);
+                  const amount = bi(this.state.amount* 10 ** 18);
```
Comment (johannbarbie) [inline src/components/Exchange.js]:
```suggestion
                  const amount = bi(this.state.amount * 10 ** 18);
```

File: src/components/Exchange.js [hunk 3]
```diff
@@ -372,24 +356,71 @@ export default class Exchange extends React.Component {
...
+        console.log("Couldn't get gas price");
+      }
     }
   }
   canSendDai() {
     return (this.state.daiSendToAddress && this.state.daiSendToAddress.length === 42 && parseFloat(this.state.daiSendAmount)>0 && parseFloat(this.state.daiSendAmount) <= parseFloat(this.props.daiBalance))
   }
 
-  async transferDai(destination,amount,message,cb) {
+  async depositDai(destination,amount,message,cb) {
```
Comment (johannbarbie) [inline src/components/Exchange.js]:
```suggestion
  async depositDai(destination, amount, message, cb) {
```

File: src/components/Exchange.js [hunk 4]
```diff
@@ -351,18 +325,28 @@ export default class Exchange extends React.Component {
...
+      changeAlert({type: 'warning',message: i18n.t('exchange.invalid_to_address')});
+    }else if(!(parseFloat(daiSendAmount) > 0)){
+      changeAlert({type: 'warning',message: i18n.t('exchange.invalid_to_amount')});
     }else{
       this.setState({
         daiToXdaiMode:"sending",
-        daiBalanceAtStart:this.props.daiBalance,
-        daiBalanceShouldBe:parseFloat(this.props.daiBalance)-parseFloat(this.state.daiSendAmount),
+        daiBalanceAtStart:daiBalance,
+        daiBalanceShouldBe:parseFloat(daiBalance)-parseFloat(daiSendAmount),
```
Comment (johannbarbie) [inline src/components/Exchange.js]:
```suggestion
        daiBalanceShouldBe:parseFloat(daiBalance) - parseFloat(daiSendAmount),
```

File: src/components/Exchange.js (hunk 1 above)
Comment (TimDaub) [reply src/components/Exchange.js]: Actually thanks for spotting this. I've been removing this `notSundai` from the code base so far. OK?

Comment (johannbarbie) [reply src/components/Exchange.js]: yes, it was only for the event. i guess if we use similar proxyToken in the future, then we will model them as assets in burner-core?

**Comment by TimDaub** [comment]
> the exit was > 1 LEAP, no-where listed. @troggy how would i start debugging the exit maker?

Actually I tried this both:

- Chrome and Metamask: https://2nuxsb25he.execute-api.eu-west-1.amazonaws.com/testnet/exits/0x51ff1fab76079d20418d1c74da65653ffe3fd0aa/0
- Chromium no Metamask: https://2nuxsb25he.execute-api.eu-west-1.amazonaws.com/testnet/exits/0x967680c79ec1ae155515bd4b923675cc9d4818dd/0

And the market maker at least received the exit. They haven't shown up here: https://testnet.leapdao.org/status but maybe it's too early.

@johannbarbie: Could it be your browser?

File: src/components/Exchange.js (hunk 1 above)
Comment (TimDaub) [reply src/components/Exchange.js]: > guess if we use similar proxyToken in the future, then we will model them as assets in burner-core?

Yup. As far as I understood this however was a problem with the market maker behaving differently for sunDai. [According to spec](https://docs.google.com/document/d/1GtS_NxhxSt9XVTphp4fWRvjeQjIVT8WbYsqpKAlWbfE/edit#heading=h.h1983kmo6k64), Exchange.js will be a separate plugin. So we can this functionality in there again when the time is right.

**Comment by johannbarbie** [comment]
> @johannbarbie: Could it be your browser?

i'll try again today and report in here: https://github.com/leapdao/burner-wallet/issues/155

**Comment by troggy** [comment]
@johannbarbie @TimDaub more input on market maker issue please as I've failed to reproduce it

https://github.com/leapdao/burner-wallet/issues/155#issuecomment-499856740

**Pull request merged**